goflags.go

```diff
@@ -342,35 +374,102 @@ func (stringSlice *StringSlice) createStringArrayDefaultValue() string {
 }
 
 func (flagSet *FlagSet) usageFunc() {
-	hashes := make(map[string]struct{})
-
 	cliOutput := flag.CommandLine.Output()
 	fmt.Fprintf(cliOutput, "%s\n\n", flagSet.description)
 	fmt.Fprintf(cliOutput, "Usage:\n  %s [flags]\n\n", os.Args[0])
 	fmt.Fprintf(cliOutput, "Flags:\n")
 
 	writer := tabwriter.NewWriter(cliOutput, 0, 0, 1, ' ', 0)
 
-	flagSet.flagKeys.forEach(func(key string, data *flagData) {
+	if len(flagSet.groups) > 0 {
+		flagSet.usageFuncForGroups(cliOutput, writer)
+	} else {
+		flagSet.usageFuncInternal(writer)
+	}
+}
+
+// usageFuncInternal prints usage for command line flags
+func (flagSet *FlagSet) usageFuncInternal(writer *tabwriter.Writer) {
+	uniqueDeduper := newUniqueDeduper()
+
+	flagSet.flagKeys.forEach(func(key string, data *FlagData) {
 		currentFlag := flag.CommandLine.Lookup(key)
 
-		dataHash := data.Hash()
-		if _, ok := hashes[dataHash]; ok {
-			return // Don't print the value if printed previously
+		if !uniqueDeduper.isUnique(data) {
+			return
 		}
-		hashes[dataHash] = struct{}{}
-
 		result := createUsageString(data, currentFlag)
 		fmt.Fprint(writer, result, "\n")
 	})
 	writer.Flush()
 }
 
+// usageFuncForGroups prints usage for command line flags with grouping enabled
+func (flagSet *FlagSet) usageFuncForGroups(cliOutput io.Writer, writer *tabwriter.Writer) {
+	uniqueDeduper := newUniqueDeduper()
+
+	var otherOptions []string
+	for _, group := range flagSet.groups {
+		fmt.Fprintf(cliOutput, "%s:\n", normalizeGroupDescription(group.description))
+
+		flagSet.flagKeys.forEach(func(key string, data *FlagData) {
```

Comment:
the difference between the `forEach` in `usageFuncForGroups` and `usageFuncInternal` is only the following: 
```go
if data.group == "" {
    otherOptions = append(otherOptions, createUsageString(data, currentFlag))
    return
}

if !strings.EqualFold(data.group, group.name) { // Ignore the flag if it's not in our intended group
    return
}
```

This means that theoretically you could extract the `forEach` logic from `usageFuncInternal` and pass in a function as a parameter that would cater for the specific content above.

Reply:
On a second thought, you can just leave it without extracting the logic, because even though it seems to be possible, it complicates the logic too much.

Reply:
yeah, that's what i thought as well!